src/votekit/ballot_graph.py

```diff
@@ -0,0 +1,226 @@
+                new_edges.append((tuple([a]+[b]+l[2:]), tuple([b]+[a]+l[2:])))
+
+        Gc.add_edges_from(new_edges)
+
+    Graphs[n] = Gc
+    return Gc
+
+
+def relabel(gr, new_label, num_cands):
+    node_map = {}
+    graph_nodes = list(gr.nodes)
+
+
+    for k in graph_nodes:
+            # add the value of new_label to every entry in every ballot
+        tmp = [new_label+y for y in k]
+
+            # reduce everything mod new_label
+        for i in range(len(tmp)):
+            if tmp[i]> num_cands:
+                tmp[i] = tmp[i]- num_cands
+        node_map[k] = tuple([new_label]+tmp)
+
+    return nx.relabel_nodes(gr, node_map)
+
+
+Graphs = {}
+
+
+class BallotGraph():
+    profile: PreferenceProfile
+    ballot_dict: dict
+    num_cands: int
+    num_voters: int
+    
+    
+    def __init__(self,profile):
+        self.profile = profile
+        self.ballot_dict= profile.to_dict()
+        
+        #cand_list = []
+        #for key in ballot_dict.keys():
+        #    for i in key:
+        #        cand_list.append(i)
+      
+        self.num_cands = len(self.profile.get_candidates())
+        
+        if self.num_cands not in Graphs.keys():
+            build_graph(self.num_cands)
+        
+        
+        all_ballots = Graphs[self.num_cands].nodes
+        di = {}
+        for ballot in all_ballots:
+            di[ballot] = 0
+        
+        self.ballot_dict = di | ballot_dict
+        
+        
+        self.clean()
+        
+        self.num_voters = sum(ballot_dict.values())
+        
+        
+    def clean(self): #deletes empty ballots, changes n-1 length ballots to n length ballots and updates counts
+        di = self.ballot_dict.copy()
+        
+        for ballot in di.keys():
+            if len(ballot)==0:
+                self.ballot_dict.pop(ballot)
+            elif len(ballot)==self.num_cands-1:
+                for i in self.profile.get_candidates():
+                    if i not in ballot:
+                        self.ballot_dict[ballot+tuple([i])]+= di[ballot]    
+                        self.ballot_dict.pop(ballot)
+                        break
+                
+        
+    def visualize(self, neighborhoods = {}):
+        Gc = Graphs[self.num_cands]    
+        if neighborhoods == {}:
+            self.clean()
+            WHITE = (1,1,1)
+            BLACK=(0,0,0)
+            cols = get_colors(self.num_cands, [WHITE,BLACK])
+            node_cols = []
+            ballots = list(Gc.nodes)
+        
+            for bal in Gc.nodes:
+                if self.ballot_dict[bal]!=0:
+                    i = self.profile.get_candidates().index(bal[0])
+                    node_cols.append(cols[i])
+                else:
+                    node_cols.append(WHITE)
+            ##want to include number of votes as part of labels,  color ballots with 0 votes grey
+        
+            nx.draw_networkx(Gc,with_labels = True, node_color = node_cols)
+            
+        else:
```

Comment:
lots of repeated code between these two conditionals, that code should be moved outside the conditionals so it isn't repeated, with only the parts that differ within the if/else blocks

Reply:
```suggestion
   def visualize(self, neighborhoods:dict = None):##visualize the whole election or select neighborhoods in the election. TODO: change this so that neighborhoods can have any neighborhood, not just heavy balls 
        Gc = build_graph(self.num_cands)  
        WHITE = (1,1,1)
        BLACK=(0,0,0)
        node_cols = []
        
        if neighborhoods==None:
            k = self.num_cands
        else:
            k = len(neighborhoods)
            
        cols = get_colors(k, [WHITE, BLACK])
        
        self.clean()
        for bal in Gc.nodes:
            i=-1
            color=WHITE
            weight = self.num_voters
            
            if neighborhoods==None:
                if self.ballot_dict[bal]!=0:
                    i=(self.profile.get_candidates()).index(bal[0])
                    
            else:
                for c in neighborhoods.keys():
                    if bal in (neighborhoods[c])[0].nodes:
                        weight = (neighborhoods[c])[1]
                        i = (list(neighborhoods.keys())).index(c)
                        break
            if i!=-1:
                color = tuple([(1-self.ballot_dict[bal]/weight)*x for x in cols[i]])
            
            node_cols.append(color)
        
        nx.draw_networkx(Gc, with_labels = True, node_color = node_cols)
        plt.show()
        return
```